sycl/include/sycl/reduction.hpp

```diff
@@ -375,43 +379,124 @@ template <typename T, class BinaryOperation, int Dims> class reducer_common {
   static constexpr int dimensions = Dims;
 };
 
+/// Templated class for common functionality of all reduction implementation
+/// classes.
+template <typename T, class BinaryOperation, bool ExplicitIdentity,
+          typename CondT = void>
+class ReductionIdentityContainer {
+  static_assert(
+      IsKnownIdentityOp<T, BinaryOperation>::value,
+      "Unspecialized case should not be hit when identity is unknown.");
```

Comment:
Stylistic nit: I'd change to

```
template <typename T, class BinaryOperation, bool ExplicitIdentity,
          typename CondT = void>
class ReductionIdentityContainer {
  static_assert(!std::is_same_v<T, T>, 
                "Partial specialization don't cover all possible options!");
};
```

and then move the current "generic" case into its own specialization.